File: LeetCode/20. Valid Parentheses/Solution.java
import java.util.Stack;

public class Solution {
    public boolean isValid(String s) {
        if (s == null) {
            return true;
        }
        char[] c = s.toCharArray();
        Stack<Character> stack = new Stack<>();
        for (int i = 0; i < c.length; i++) {
            switch (c[i]) {
                case '(':
                case '[':
                case '{':
                    stack.push(c[i]);
                    break;
                case ')':
                    if (stack.empty() || stack.pop() != '(') {
                        return false;
                    }
                    break;
                case ']':
                    if (stack.empty() || stack.pop() != '[') {
                        return false;
                    }
                    break;
                case '}':
                    if (stack.empty() || stack.pop() != '{') {
                        return false;
                    }
                    break;
                default:

            }
        }
        return stack.empty();
    }
}

File: LeetCode/20. Valid Parentheses/solution.py
from collections import deque


class Solution(object):
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        stack = deque()
        left, right = '([{', ')]}'
        for ch in s:
            if ch in left:
                stack.append(ch)
            else:
                index = right.index(ch)
                if not stack or stack[-1] != left[index]:
                    return False
                else:
                    stack.pop()
                    
        return len(stack) == 0
